src/mempool/mempool.rs

```diff
@@ -116,6 +116,139 @@ pub struct MempoolMessagesRequest {
     pub max_messages_per_block: u32,
 }
 
+fn message_already_exists(
+    shard: u32,
+    shard_stores: &mut HashMap<u32, Stores>,
+    message: &MempoolMessage,
+) -> bool {
+    let fid = message.fid();
+
+    let stores = shard_stores.get_mut(&shard);
+    // Default to false in the error paths
+    match stores {
+        None => {
+            error!("Error finding store for shard: {}", shard);
+            false
+        }
+        Some(stores) => match message {
+            MempoolMessage::UserMessage(message) => match &message.data {
+                None => false,
+                Some(message_data) => {
+                    let ts_hash = make_ts_hash(message_data.timestamp, &message.hash).unwrap();
+                    match type_to_set_postfix(message_data.r#type()) {
+                        Err(err) => {
+                            error!("Error retrieving set postfix: {}", err.to_string());
+                            false
+                        }
+                        Ok(set_postfix) => {
+                            let primary_key =
+                                make_message_primary_key(fid, set_postfix as u8, Some(&ts_hash));
+                            let existing_message = get_message_by_key(
+                                &stores.db,
+                                &mut RocksDbTransactionBatch::new(),
+                                &primary_key,
+                            );
+                            match existing_message {
+                                Ok(Some(_)) => true,
+                                Err(_) | Ok(None) => false,
+                            }
+                        }
+                    }
+                }
+            },
+            MempoolMessage::ValidatorMessage(message) => {
+                if let Some(onchain_event) = &message.on_chain_event {
+                    match stores.onchain_event_store.exists(&onchain_event) {
+                        Err(_) => return false,
+                        Ok(exists) => return exists,
+                    }
+                }
+
+                if let Some(fname_transfer) = &message.fname_transfer {
+                    match &fname_transfer.proof {
+                        None => return false,
+                        Some(proof) => {
+                            let username_proof = UserDataStore::get_username_proof(
+                                &stores.user_data_store,
+                                &mut RocksDbTransactionBatch::new(),
+                                &proof.name,
+                            );
+                            match username_proof {
+                                Err(_) | Ok(None) => return false,
+                                Ok(Some(_)) => return true,
+                            }
+                        }
+                    }
+                }
+                return false;
+            }
+        },
+    }
+}
+
+async fn gossip_message(
```

Comment:
I think generally it's better to favor composition (esp. in rust) over small context-less functions. This is small, so not a big deal. 

You could do something like a ReadOnlyMempool which has the validation and gossip features, and then have the ValidatorMempool hold a reference to it and delegate to it where required. Or create a separate struct that holds the common logic that's used by both the ReadNodeMempool and the ValidatorMempool

There are going to be other checks we'll need to add down the line like rate limits and validation (I think we need to do the validation right now as well).